### PR TITLE
Add a dummy actions template to enable it on CI

### PR DESCRIPTION
**Pull request by lzutao**
I already had a complete template to transition from Travis and Appveyor.
Since actions will not run on repo haven't enabled it before.
I have this dummy template first to enable it on default branch.
We should address a few todo here, merge it and I will send another patch
to complete the migration.

### CODE REVIEW

**Comment by RalfJung** [comment]
So the plan is to migrate Miri to GH actions? That's okay I guess, I am just confused that you act as if everyone would already know that.^^ Cc @rust-lang/miri 

Bors configuration will then likely also have to be changed.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    branches:
+      - 'auto'
+      - 'try'
+      - '!gh-pages'
+  pull_request:
+    branches:
+      - 'master'
+  # TODO: fill the cron here
+  # schedule:
+  #   - cron: "30 0 * * 1" # Every Monday at half past midnight UTC skip-pr skip-stable
```
Comment (RalfJung) [inline .github/workflows/ci.yml:17]: Miri wants a daily cronjob, not just a weekly one.

File: README.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
-# Miri [![Build Status](https://travis-ci.com/rust-lang/miri.svg?branch=master)](https://travis-ci.com/rust-lang/miri) [![Windows build status](https://ci.appveyor.com/api/projects/status/github/rust-lang/miri?svg=true)](https://ci.appveyor.com/project/rust-lang-libs/miri)
+# Miri
 
+[![Build Status][actions-badge]][actions-url]
+
+[actions-badge]: https://github.com/rust-lang/miri/workflows/CI/badge.svg?branch=master
+[actions-url]: https://github.com/rust-lang/miri/actions
```
Comment (RalfJung) [inline README.md:6]: Let's only change badges when we actually change the main CI.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            os: ubuntu-latest
+            target: x86_64-unknown-linux-gnu
+          - build: macos
+            os: macos-latest
+            target: x86_64-apple-darwin
+          - build: win32
+            os: windows-latest
+            target: i686-pc-windows-msvc
+    steps:
+      - uses: actions/checkout@v2
```
Comment (RalfJung) [inline .github/workflows/ci.yml:39]: So, uh, where does this define the actual commands that get run?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  pull_request:
+    branches:
+      - 'master'
+  # TODO: fill the cron here
+  # schedule:
+  #   - cron: "30 0 * * 1" # Every Monday at half past midnight UTC skip-pr skip-stable
+
+jobs:
+  build:
+    runs-on: ${{ matrix.os }}
```
Comment (RalfJung) [inline .github/workflows/ci.yml:21]: Do you know a good introduction to GH action syntax? For example, where is this `matrix.os` variable defined?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    env:
+      RUST_BACKTRACE: 1
+      PROJECT_NAME: miri
```
Comment (RalfJung) [inline .github/workflows/ci.yml:24]: What is this environment variable for?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+name: CI
+
+# About email notifications, see <https://git.io/JUo27>.
```
Comment (RalfJung) [inline .github/workflows/ci.yml:3]: That is a long document covering GH email notifications in general. I don't understand what this is supposed to tell me here.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (bjorn3) [reply .github/workflows/ci.yml:21]: It is defined by the `matrix` starting from line 26 to 37.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (lzutao) [reply .github/workflows/ci.yml:24]: from appveyor template. I should have check that it is unused.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (lzutao) [reply .github/workflows/ci.yml:17]: But what is the specific time ?

File: .github/workflows/ci.yml (hunk 6 above)
Comment (lzutao) [reply .github/workflows/ci.yml:3]: like this part in travis.yml
https://github.com/rust-lang/miri/blob/aa832c1133217a079ae5d9af65dd0b7ad96aca07/.travis.yml#L52-L57

In short, you receive mails for build failure.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (RalfJung) [reply .github/workflows/ci.yml:17]: It doesn't really matter... let's say 7:00 UTC?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (lzutao) [reply .github/workflows/ci.yml:39]: This uses a predefined template used to checkout the build repository.
https://github.com/actions/checkout

File: .github/workflows/ci.yml (hunk 6 above)
Comment (RalfJung) [reply .github/workflows/ci.yml:3]: We mostly rely on bors for this nowadays (these notifications predate us using bors)... so I'd say let's go with whatever a reasonable default is and we can always tweak it later.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (RalfJung) [reply .github/workflows/ci.yml:39]: Oh I see, so while this happens by default on other CI runners, it needs to be specifically included for GH? And no actual tests are run yet? Makes sense.

**Comment by lzutao** [comment]
> I am just confused that you act as if everyone would already know that.

I'm sorry. I suddenly have the idea to transition Miri to github actions. Hope it not a bad idea.

> Bors configuration will then likely also have to be changed.

Yeah, someone in infra team can do that.

> Do you know a good introduction to GH action syntax?

I don't. The official demonstration example of Rust is <https://github.com/actions/starter-workflows/blob/main/ci/rust.yml>.
But there is official documentation to catch up:
- https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions
- https://docs.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners
- https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions
- https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable
- https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions
- https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables

File: .github/workflows/ci.yml (hunk 6 above)
Comment (lzutao) [reply .github/workflows/ci.yml]: I'll remove it then.

**Comment by RalfJung** [comment]
> I'm sorry. I suddenly have the idea to transition Miri to github actions. Hope it not a bad idea.

Funny idea for a Sunday morning :rofl: . But sure, since this seems to be what the Rust GH org is moving to in general, that is fine for me. Thanks for taking initiative. :)
I am not sure if we have to coordinate with the infra team for this... @pietroalbini?

> Yeah, someone in infra team can do that.

AFAIK anyone can, it's a PR to change [this file](https://github.com/rust-lang/rust-central-station/blob/1b21ad5adf9204f7867024a3a56c9584d6359581/homu.toml.template#L380-L383).

Thanks for the docs link, looks like I have some reading to do. :)

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -12,7 +12,7 @@ on:
       - 'master'
   schedule:
     # Use <https://crontab.guru> to conveniently edit cron schedule.
-    - cron: "0 7 * * *" "30 0 * * 1" # At 07:00 everyday
+    - cron: "0 7 * * *" # At 07:00 everyday
```
Comment (RalfJung) [inline .github/workflows/ci.yml:15]:
```suggestion
    - cron: "0 7 * * *" # At 07:00 every day.
```

Comment (lzutao) [reply .github/workflows/ci.yml]: According to manpage

> The CRON_TZ variable specifies the time zone specific for the cron table

But it seems like github runners choose GMT time: https://github.com/actions/virtual-environments/blob/f199e4ad671a97e32b05aec02e87b6c6404caa17/images/macos/provision/configuration/configure-machine.sh#L6-L9

**Comment by lzutao** [comment]
> AFAIK anyone can, it's a PR to change [this file](https://github.com/rust-lang/rust-central-station/blob/1b21ad5adf9204f7867024a3a56c9584d6359581/homu.toml.template#L380-L383).

Nice, never known this before.

**Comment by pietroalbini** [comment]
> I am not sure if we have to coordinate with the infra team for this...

Not at all, every project is free to use whatever CI system they prefer. GHA is probably the best as we have a lot of capacity there, but y'all can use or switch to whatever you want :)

**Comment by RalfJung** [comment]
@pietroalbini great. :) Capacity would have been my next question... so far we only ran Linux and macOS tests in PRs, and Windows tests only in bors. We could still do that if you'd prefer to to reduce CI load a bit.

**Comment by lzutao** [comment]
> GHA is probably the best as we have a lot of capacity there

I thought the capacity only for rust-lang/rust, so it is available for all projects under rust-lang org.

**Comment by pietroalbini** [comment]
> great. :) Capacity would have been my next question... so far we only ran Linux and macOS tests in PRs, and Windows tests only in bors. We could still do that if you'd prefer to to reduce CI load a bit.

Running Windows on PRs too sounds fine, three builders is not a problem.

> I thought the capacity only for rust-lang/rust, so it is available for all projects under rust-lang org.

The XL runners should be used only by rust-lang/rust, but we had the concurrency increased for the whole org.

**Comment by lzutao** [comment]
This PR will be merged before https://github.com/rust-lang/rust-central-station/pull/939

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,36 @@
+name: CI
+
+on:
+  push:
+    # Run in PRs and for bors, but not on master.
+    branches:
+      - 'auto'
+      - 'try'
+      - '!gh-pages'
```
Comment (RalfJung) [inline .github/workflows/ci.yml:9]: Why is this needed?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    - cron: "0 7 * * *" # At 07:00 UTC every day.
+
+jobs:
+  build:
+    runs-on: ${{ matrix.os }}
+    env:
+      RUST_BACKTRACE: 1
+    strategy:
+      matrix:
+        build: [linux64, macos, win32]
```
Comment (RalfJung) [inline .github/workflows/ci.yml:24]: Couldn't you use `os: [ubuntu-latest, macos-latest, windows-latest]` here and do away with the `build` variable entirely?

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        include:
+          - build: linux64
+            os: ubuntu-latest
+            target: x86_64-unknown-linux-gnu
```
Comment (RalfJung) [inline .github/workflows/ci.yml:28]: Since we do cross-testing, it might be better to call this `host_target`.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (lzutao) [reply .github/workflows/ci.yml:24]: You mean replacing `build` with `os`?
Could be if you want combination build jobs: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-including-additional-values-into-combinations

File: .github/workflows/ci.yml (hunk 8 above)
Comment (lzutao) [reply .github/workflows/ci.yml:9]: redundant when copied from other templates where every branches allowed but gh-page, will remove.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (lzutao) [reply .github/workflows/ci.yml:28]: for win32, it isn't host_target, because the builder is x86_64.
Don't worry about this, since I will use OS name instead of target triple for cross-testing.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (RalfJung) [reply .github/workflows/ci.yml:23]: Not replacing it obviously, also changing the value.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (RalfJung) [reply .github/workflows/ci.yml:27]: > it isn't host_target, because the builder is x86_64.

It should be the host target though! By "host target" I mean the target of the host rustc that gets installed. Even if this is an x86_64 builder, we want to install the 32bit rustc. So, the host-target should be a 32bit target still.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (lzutao) [reply .github/workflows/ci.yml:23]: Could be, but those value in `build` also are the names when displayed in github CI checks.
We might prefer win32 over windows-latest because in the future, 
we might add win64 or linux32 build.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (lzutao) [reply .github/workflows/ci.yml]: if you say so.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (RalfJung) [reply .github/workflows/ci.yml:23]: I see... Is this the usual way to structure these matrices?

Comment (lzutao) [reply .github/workflows/ci.yml:23]: I often see alex projects doing like that: 
https://github.com/alexcrichton/cc-rs/blob/master/.github/workflows/main.yml
https://github.com/alexcrichton/tar-rs/blob/master/.github/workflows/main.yml
https://github.com/alexcrichton/xz2-rs/blob/master/.github/workflows/main.yml

Comment (RalfJung) [reply .github/workflows/ci.yml:23]: I see, makes sense. Thanks!

**Comment by RalfJung** [comment]
Looks like a good start, thanks :)
@bors r+
